src/services/admin/AuditLogsService.ts

```diff
@@ -0,0 +1,500 @@
+              publishedChanges
+                .concat(auditLogs)
+                .sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime())
+            )
+          )
+      )
+  }
+
+  getAuditLogsViaFormsg(
+    email: string,
+    repoNames: string[],
+    sinceDate: string,
+    untilDate: string,
+    formSubmissionId: string
+  ): ResultAsync<void, AuditLogsError | ForbiddenError> {
+    const emailSubject = `[Isomer] Site Audit Logs (submission ID: ${formSubmissionId})`
+    logger.info(
+      `Received request to get audit logs, submission ID: ${formSubmissionId}`
+    )
+
+    // Step 1: Check if the user exists
+    return ResultAsync.fromPromise(this.usersService.findByEmail(email), () => {
+      logger.error(
+        `Error occurred while retrieving user data from the database for email ${email}`
+      )
+
+      return new DatabaseError(
+        "Error occurred while retrieving user data from the database"
+      )
+    })
+      .andThen((user) => {
+        if (!user) {
+          logger.warn(`Email address ${email} is not registered on Isomer CMS`)
+
+          return errAsync(
+            new ForbiddenError("Email address is not registered on Isomer CMS")
+          )
+        }
+
+        return okAsync(user)
+      })
+      .andThen((user) =>
+        // Step 2: Check if the user is a collaborator of ALL given repos
+        ResultAsync.combine([
+          okAsync(user),
+          ...repoNames.map((repoName) =>
+            ResultAsync.fromPromise(
+              this.collaboratorsService.getRole(repoName, user.id.toString()),
+              () =>
+                new DatabaseError(
+                  `Error retrieving user's role for the repo ${repoName}`
+                )
+            )
+          ),
+        ])
+      )
+      .andThen(([user, ...roles]) => {
+        const isUserNotAdminCollaborator = roles.some(
+          (role) =>
+            !role ||
+            (role !== CollaboratorRoles.Admin &&
+              role !== CollaboratorRoles.IsomerAdmin)
+        )
+
+        if (isUserNotAdminCollaborator) {
+          logger.warn(
+            `User ${email} is not an admin collaborator of all the given repos ${JSON.stringify(
+              repoNames
+            )}`
+          )
+
+          return errAsync(
+            new ForbiddenError(
+              "User is not an admin collaborator of all the given repos"
+            )
+          )
+        }
+
+        return okAsync(user)
+      })
+      .andThen((user) =>
+        // Step 3: Obtain a token for the user
+        ResultAsync.combine([
+          okAsync(user),
+          tokenServiceInstance.getAccessToken(),
+        ])
+      )
+      .andThen(([user, accessToken]) =>
+        // Step 4: Construct a fake user session data object and get audit logs
+        ResultAsync.combine(
+          repoNames.map((repoName) => {
+            const userSessionData = new UserWithSiteSessionData({
+              githubId: user.githubId,
```

Comment:
what if this doesnt exist? ie an admin that uses this form that only has email sites?

Reply:
It would be set to undefined, but that is not used anyway since we rely on the accessToken from TokenService.

Reply:
this is confusing to me :(
if it is not needed here, why declare it? doesnt seem like a necessary key too

Reply:
My bad, it is defined and is needed because of this control flow:

https://github.com/isomerpages/isomercms-backend/blob/24658ec51af72594c7da2c6368ebc0f7be5d5c52/src/classes/UserSessionData.ts#L32-L35

Otherwise, it would not be possible for us to set the accessToken for the request.

Reply:
wait this is an issue right? not all users have a github id, wouldnt this then fail for those users?

Reply:
Adjusted in ae8410a1d315b913370a8333c2427fd174c9c067.